src/core/geometry/qgscurvepolygon.cpp

```diff
@@ -492,6 +492,70 @@ double QgsCurvePolygon::area() const
   return totalArea;
 }
 
+double QgsCurvePolygon::area3D() const
+{
+  double result = 0.0;
+
+  if ( !mExteriorRing )
+  {
+    return result;
+  }
+
+  QgsPoint ptA;
+  QgsPoint ptB;
+  QgsPoint ptC;
+
+  if ( !hasPlane3D( ptA, ptB, ptC ) )
+  {
+    return result;
+  }
+
+  // Build an orthonormal reference frame (ux, uy, uz) from three 3D points
+  QgsVector3D ux( ptC.x() - ptB.x(), ptC.y() - ptB.y(), ptC.z() - ptB.z() );
+  QgsVector3D uz = QgsVector3D::crossProduct( ux, QgsVector3D( ptA.x() - ptB.x(), ptA.y() - ptB.y(), ptA.z() - ptB.z() ) );
+  ux = ux / std::sqrt( ux.lengthSquared() );
+  uz = uz / std::sqrt( uz.lengthSquared() );
+  const QgsVector3D uy = QgsVector3D::crossProduct( uz, ux );
+
+  /*
+   * compute the area for each ring in the local basis
+   */
+  const int numRings = static_cast<int>( mInteriorRings.size() + 1 );
+  for ( int ringIndex = 0; ringIndex < numRings; ++ringIndex )
+  {
+    const QgsLineString *ring = static_cast< const QgsLineString * >( ringIndex == 0 ? mExteriorRing.get() : mInteriorRings.at( ringIndex - 1 ) );
```

Comment:
This should use qgsgeometry_cast, but more importantly -- it will crash for curved geometry types!